File: src/server/GuideAntsApi.DataModel/Models/HostFolderMount.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace GuideAntsApi.DataModel.Models;

[Index(nameof(ProjectId), nameof(Status))]
[Index(nameof(MountKey), IsUnique = true)]
public class HostFolderMount
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    public Guid ProjectId { get; set; }

    public Project Project { get; set; } = null!;

    public Guid? NotebookId { get; set; }

    public Notebook? Notebook { get; set; }

    [Required]
    public HostFolderMountScope Scope { get; set; }

    [Required]
    public SourceKind SourceKind { get; set; }

    [Required]
    [StringLength(255)]
    public string DisplayName { get; set; } = string.Empty;

    [Required]
    [StringLength(255)]
    public string LeafName { get; set; } = string.Empty;

    [Required]
    [StringLength(128)]
    public string MountKey { get; set; } = string.Empty;

    /// <summary>
    /// LocalPath: absolute host path. Smb: display/UNC form. Admin-only sensitive.
    /// </summary>
    [Required]
    [AdminOnlySensitive]
    public string SourceSpec { get; set; } = string.Empty;

    [Required]
    [StringLength(512)]
    public string ContainerSourcePath { get; set; } = string.Empty;

    /// <summary>Smb only: CIFS device (e.g. //server/share/sub). Null for LocalPath.</summary>
    [StringLength(1024)]
    public string? NetworkDevice { get; set; }

    /// <summary>Smb only: non-secret mount options (uid, gid, file_mode, etc.). Null for LocalPath.</summary>
    [StringLength(2048)]
    public string? NetworkOptions { get; set; }

    /// <summary>Smb only: secret-store reference — never raw credentials. Admin-only sensitive.</summary>
    [StringLength(512)]
    [AdminOnlySensitive]
    public string? CredentialRef { get; set; }

    [Required]
    public HostFolderMountStatus Status { get; set; } = HostFolderMountStatus.PendingRestart;

    [Required]
    public Guid CreatedByUserId { get; set; }

    public User CreatedByUser { get; set; } = null!;

    [Required]
    public DateTime CreatedUtc { get; set; } = DateTime.UtcNow;

    [Required]
    public DateTime UpdatedUtc { get; set; } = DateTime.UtcNow;

    public DateTime? RemovedUtc { get; set; }

    [StringLength(4000)]
    public string? ErrorMessage { get; set; }

    public ICollection<HostFolderMountLink> Links { get; set; } = [];
}

File: src/client/src/types/hostFolderMount.ts
export type HostFolderMountScope = 'Notebook' | 'Project';

export type HostFolderMountStatus =
  | 'PendingRestart'
  | 'Active'
  | 'PendingRemoval'
  | 'Removed'
  | 'Error';

export type HostFolderMountLinkStatus =
  | 'PendingRestart'
  | 'PendingLink'
  | 'Linked'
  | 'Unlinked'
  | 'LinkError'
  | 'UnlinkError';

export type HostFolderMountDisplayState =
  | 'PendingRestart'
  | 'Linked'
  | 'MissingSource'
  | 'LinkError'
  | 'PendingRemoval';

export interface HostFolderMountSummaryDto {
  mountId: string;
  projectId: string;
  notebookId: string | null;
  scope: HostFolderMountScope;
  sourceKind: string;
  displayName: string;
  leafName: string;
  mountKey: string;
  containerSourcePath: string;
  status: HostFolderMountStatus;
  createdUtc: string;
  updatedUtc: string;
  errorMessage: string | null;
}

export interface HostFolderMountLinkSummaryDto {
  linkId: string;
  notebookId: string;
  linkRelativePath: string;
  status: HostFolderMountLinkStatus;
  errorMessage: string | null;
}

export interface HostFolderMountDetailDto extends HostFolderMountSummaryDto {
  removedUtc: string | null;
  hostPath: string | null;
  networkDevice: string | null;
  networkOptions: string | null;
  credentialRef: string | null;
  links: HostFolderMountLinkSummaryDto[];
}

export interface HostFolderMountCreateRequest {
  notebookId?: string;
  scope: HostFolderMountScope;
  hostPath: string;
  leafName?: string;
}

export interface HostFolderMountCreateResponse {
  mountId: string;
  status: HostFolderMountStatus;
  leafName: string;
  containerSourcePath: string;
  command: string;
}

export interface HostFolderMountCommandResponse {
  mountId: string;
  status: HostFolderMountStatus;
  command: string;
  leafName?: string;
  containerSourcePath?: string;
}

export interface HostFolderMountReconcileResponse {
  mountId: string;
  status: HostFolderMountStatus;
  message: string;
}

export interface NotebookHostMountEntry {
  mountId: string;
  leafName: string;
  relativePath: string;
  displayName: string;
  displayState: HostFolderMountDisplayState;
  mountStatus: HostFolderMountStatus;
  scope: HostFolderMountScope;
  linkStatus: HostFolderMountLinkStatus | null;
}

export interface ProjectHostMountEntry {
  mountId: string;
  leafName: string;
  displayName: string;
  displayState: HostFolderMountDisplayState;
  mountStatus: HostFolderMountStatus;
}
